File: python/benchmarks/bench_lru_cache.py
#!/usr/bin/env python3
"""LRU cache benchmark: put_miss, put_hit, get_hit, get_miss, eviction (same as Rust)."""

import sys
import os
import random
import time

_script_dir = os.path.dirname(os.path.abspath(__file__))
sys.path.insert(0, _script_dir)
sys.path.insert(0, os.path.join(_script_dir, ".."))
from src.lru_cache import LRUCache
from bench_common import (
    SCALES,
    NUM_RUNS,
    get_memory_mb,
    mean_std,
    write_csv,
    get_results_dir,
)


def main():
    results_dir = get_results_dir()
    os.makedirs(results_dir, exist_ok=True)

    rows = [
        [
            "N",
            "put_miss_mean_ms",
            "put_miss_std_ms",
            "put_hit_mean_ms",
            "put_hit_std_ms",
            "get_hit_mean_ms",
            "get_hit_std_ms",
            "get_miss_mean_ms",
            "get_miss_std_ms",
            "eviction_mean_ms",
            "eviction_std_ms",
            "memory_mb",
        ]
    ]
    for N in SCALES:
        keys = list(range(N))
        capacity = max(16, N)

        # Warm-up: build and use cache once at this scale
        cache = LRUCache(capacity)
        for k in keys:
            cache.put(k, k)
        for k in keys:
            _ = cache.get(k)

        put_miss_times = []
        put_hit_times = []
        get_hit_times = []
        get_miss_times = []
        eviction_times = []

        for _ in range(NUM_RUNS):
            # put_miss: cache empty, then (capacity-1) puts of new keys (no eviction)
            cache = LRUCache(capacity)
            start = time.perf_counter()
            for i in range(capacity - 1):
                cache.put(i, i)
            put_miss_times.append((time.perf_counter() - start) * 1000)

            # put_hit: full cache, N updates of existing keys
            cache = LRUCache(capacity)
            for k in keys:
                cache.put(k, k)
            start = time.perf_counter()
            for i in range(N):
                cache.put(i % capacity, i)
            put_hit_times.append((time.perf_counter() - start) * 1000)

            # get_hit: full cache, N lookups of existing keys
            cache = LRUCache(capacity)
            for k in keys:
                cache.put(k, k)
            start = time.perf_counter()
            for i in range(N):
                _ = cache.get(i % capacity)
            get_hit_times.append((time.perf_counter() - start) * 1000)

            # get_miss: full cache, N lookups of a key not in cache
            cache = LRUCache(capacity)
            for k in keys:
                cache.put(k, k)
            missing = N  # not in 0..N
            start = time.perf_counter()
            for _ in range(N):
                _ = cache.get(missing)
            get_miss_times.append((time.perf_counter() - start) * 1000)

            # eviction: full cache, N puts of new keys so each put evicts LRU
            cache = LRUCache(capacity)
            for k in keys:
                cache.put(k, k)
            start = time.perf_counter()
            for i in range(N, 2 * N):
                cache.put(i, i)
            eviction_times.append((time.perf_counter() - start) * 1000)

        pm_mean, pm_std = mean_std(put_miss_times)
        ph_mean, ph_std = mean_std(put_hit_times)
        gh_mean, gh_std = mean_std(get_hit_times)
        gm_mean, gm_std = mean_std(get_miss_times)
        ev_mean, ev_std = mean_std(eviction_times)
        memory_mb = get_memory_mb()
        rows.append(
            [
                N,
                f"{pm_mean:.6f}",
                f"{pm_std:.6f}",
                f"{ph_mean:.6f}",
                f"{ph_std:.6f}",
                f"{gh_mean:.6f}",
                f"{gh_std:.6f}",
                f"{gm_mean:.6f}",
                f"{gm_std:.6f}",
                f"{ev_mean:.6f}",
                f"{ev_std:.6f}",
                f"{memory_mb:.4f}",
            ]
        )
        print(
            f"N={N}: put_miss {pm_mean:.6f}±{pm_std:.6f} ms, put_hit {ph_mean:.6f}±{ph_std:.6f} ms, "
            f"get_hit {gh_mean:.6f}±{gh_std:.6f} ms, get_miss {gm_mean:.6f}±{gm_std:.6f} ms, "
            f"eviction {ev_mean:.6f}±{ev_std:.6f} ms, memory={memory_mb:.4f} MB"
        )
    write_csv(rows, results_dir, "python_lru_cache.csv")


if __name__ == "__main__":
    main()

File: cpp/benchmarks/bench_lru_cache.cpp
#include <chrono>
#include <random>

#include "bench_common.hpp"
#include "lru_cache.hpp"

int main() {
    std::string out_dir = get_results_dir();
    fs::create_directories(out_dir);

    std::random_device rd;
    std::mt19937 g(rd());

    std::string csv_path = out_dir + "/cpp_lru_cache.csv";
    std::ofstream file(csv_path);
    if (!file.is_open()) {
        std::cerr << "Failed to open " << csv_path << "\n";
        return 1;
    }
    file << "N,put_miss_mean_ms,put_miss_std_ms,put_hit_mean_ms,put_hit_std_ms,get_hit_mean_ms,"
            "get_hit_std_ms,get_miss_mean_ms,get_miss_std_ms,eviction_mean_ms,eviction_std_ms,"
            "memory_mb\n";
    file << std::fixed << std::setprecision(6);

    for (int n : SCALES) {
        std::vector<int> keys(n);
        std::iota(keys.begin(), keys.end(), 0);
        std::shuffle(keys.begin(), keys.end(), g);

        size_t capacity = std::max(size_t(16), static_cast<size_t>(n));
        {  // Warm-up: build and use cache once at this scale
            lru_cache::LRUCache cache(capacity);
            for (int k : keys) cache.put(k, k);
            for (int k : keys) do_not_optimize(cache.get(k));
        }

        std::vector<double> put_miss_ms(NUM_RUNS), put_hit_ms(NUM_RUNS), get_hit_ms(NUM_RUNS),
            get_miss_ms(NUM_RUNS), eviction_ms(NUM_RUNS);

        for (int run = 0; run < NUM_RUNS; run++) {
            // put_miss: cache empty, then (capacity-1) puts of new keys (no eviction)
            {
                lru_cache::LRUCache cache(capacity);
                auto start = std::chrono::high_resolution_clock::now();
                for (size_t i = 0; i < capacity - 1; i++)
                    cache.put(static_cast<int>(i), static_cast<int>(i));
                put_miss_ms[run] = std::chrono::duration<double, std::milli>(
                                       std::chrono::high_resolution_clock::now() - start)
                                       .count();
            }

            // put_hit: full cache, n updates of existing keys
            {
                lru_cache::LRUCache cache(capacity);
                for (int k : keys) cache.put(k, k);
                auto start = std::chrono::high_resolution_clock::now();
                for (int i = 0; i < n; i++) cache.put(i % static_cast<int>(capacity), i);
                put_hit_ms[run] = std::chrono::duration<double, std::milli>(
                                      std::chrono::high_resolution_clock::now() - start)
                                      .count();
            }

            // get_hit: full cache, n lookups of existing keys
            {
                lru_cache::LRUCache cache(capacity);
                for (int k : keys) cache.put(k, k);
                auto start = std::chrono::high_resolution_clock::now();
                for (int i = 0; i < n; i++)
                    do_not_optimize(cache.get(i % static_cast<int>(capacity)));
                get_hit_ms[run] = std::chrono::duration<double, std::milli>(
                                      std::chrono::high_resolution_clock::now() - start)
                                      .count();
            }

            // get_miss: full cache, n lookups of a key not in cache
            {
                lru_cache::LRUCache cache(capacity);
                for (int k : keys) cache.put(k, k);
                const int missing = n;
                auto start = std::chrono::high_resolution_clock::now();
                for (int i = 0; i < n; i++) do_not_optimize(cache.get(missing));
                get_miss_ms[run] = std::chrono::duration<double, std::milli>(
                                       std::chrono::high_resolution_clock::now() - start)
                                       .count();
            }

            // eviction: full cache, n puts of new keys so each put evicts LRU
            {
                lru_cache::LRUCache cache(capacity);
                for (int k : keys) cache.put(k, k);
                auto start = std::chrono::high_resolution_clock::now();
                for (int i = n; i < 2 * n; i++) cache.put(i, i);
                eviction_ms[run] = std::chrono::duration<double, std::milli>(
                                       std::chrono::high_resolution_clock::now() - start)
                                       .count();
            }
        }

        double pm_mean, pm_std, ph_mean, ph_std, gh_mean, gh_std, gm_mean, gm_std, ev_mean, ev_std;
        mean_std(put_miss_ms, pm_mean, pm_std);
        mean_std(put_hit_ms, ph_mean, ph_std);
        mean_std(get_hit_ms, gh_mean, gh_std);
        mean_std(get_miss_ms, gm_mean, gm_std);
        mean_std(eviction_ms, ev_mean, ev_std);
        double mem = memory_mb();
        file << n << "," << pm_mean << "," << pm_std << "," << ph_mean << "," << ph_std << ","
             << gh_mean << "," << gh_std << "," << gm_mean << "," << gm_std << "," << ev_mean << ","
             << ev_std << "," << std::setprecision(4) << mem << "\n";
        file << std::setprecision(6);
        std::cout << "N=" << n << ": put_miss " << pm_mean << "±" << pm_std << " ms, put_hit "
                  << ph_mean << "±" << ph_std << " ms, get_hit " << gh_mean << "±" << gh_std
                  << " ms, get_miss " << gm_mean << "±" << gm_std << " ms, eviction " << ev_mean
                  << "±" << ev_std << " ms, memory=" << mem << " MB\n";
    }
    std::cout << "Wrote " << csv_path << "\n";
    return 0;
}
